### PR TITLE
Lick interval update

### PR DESCRIPTION
**Pull request by micahwoodard**
### Pull Request instructions:
- Please follow the [update protocol](https://github.com/AllenNeuralDynamics/aind-behavior-blog/wiki/Software-Update-Procedures)
- Answer the questions below in detail. Your responses will be emailed to experimenters. 
- If the experimenters must do anything new, provide detailed step by step instructions on the [wiki](https://github.com/AllenNeuralDynamics/aind-behavior-blog/wiki)
- If computer maintainers need to manually update anything, provide detailed step by step instructions
- Use markdown syntax in order for your comments to be rendered reliably in the email: "1." instead of "1)", use four spaces for indents.
- If you use the keyword "skip email" in the title, it will skip the email updates
- Merges from "develop" into "production_testing" should use the keyword "production merge" in the title for reliable indexing of updates
- Merges from "production_testing" into "main" should use the keyword "update main"
  
### Describe changes:
- updates user if percentage of same and cross side licks under 100 ms is larger than 10%
### What issues or discussions does this update address?
- resolves [#308](https://github.com/AllenNeuralDynamics/aind-behavior-blog/issues/308)
### Describe the expected change in behavior from the perspective of the experimenter
- text boxes in the lick distribution dialog box will display warning if percentage of same and cross side licks under 100 ms is larger than 10%
### Describe any manual update steps for task computers
- N/A
### Was this update tested in 446/447?
- [x] 446/447

### CODE REVIEW

**Comment by micahwoodard** [comment]
I've tested what I could in the UI and then tested the calculate function outside, but I am unsure how to test without a real licks? Using the simulated mouse doesn't seem to use the GenerateTrials but I could be wrong and am not seeing it. Any ideas would be great!

**Comment by micahwoodard** [comment]
- load old data and trigger function
- ask Ella if there is a mouse to test on

File: src/foraging_gui/MyFunctions.py
Comment (alexpiet) [inline src/foraging_gui/MyFunctions.py:910]: have you tried the edge case of just one right, or just one left lick?

File: src/foraging_gui/MyFunctions.py
Comment (alexpiet) [inline src/foraging_gui/MyFunctions.py:913]: On line 913 you can considering clarify this code:
```
same_side_l_frac = len(same_side_l[~(same_side_l > threshold)]) / len(same_side_l)
```
into this
```
same_side_l_frac = np.mean(same_side_l  <= threshold)
```

Notes:
- `~(same_side_l > threshold)` is the same as `same_side_l <= threshold`
- You can get the count of a boolean array with `np.sum(same_side_l <=threshold)`, and then divide by `len(same_side_l)`
- You can get the % true of a boolean array with `np.mean(same_side_l <= threshold)`

File: src/foraging_gui/MyFunctions.py
Comment (alexpiet) [inline src/foraging_gui/MyFunctions.py:923]: I'm not sure if adding these fractions is what we want. 
I think if either the left or right is above .1 (which you should set as a variable instead of hard coding), then we should alert the user

File: src/foraging_gui/MyFunctions.py
Comment (alexpiet) [inline src/foraging_gui/MyFunctions.py:933]: I'm not sure I understand what `merged_sorted` contains

File: src/foraging_gui/MyFunctions.py
Comment (alexpiet) [inline src/foraging_gui/MyFunctions.py:940]: This code block is very confusing, and I'm not totally sure whats going on. 

I think I would have implemented this with a while loop and a pointer in each of the left and right lick times

File: src/foraging_gui/MyFunctions.py
Comment (micahwoodard) [reply src/foraging_gui/MyFunctions.py:923]: Okay I was doing that based on what you said [here](https://github.com/orgs/AllenNeuralDynamics/projects/47/views/10?sortedBy%5Bdirection%5D=asc&sortedBy%5BcolumnId%5D=Assignees&sortedBy%5Bdirection%5D=desc&sortedBy%5BcolumnId%5D=Status&pane=issue&itemId=69427925&issue=AllenNeuralDynamics%7Caind-behavior-blog%7C308#:~:text=If%20the%20proportion%20of%20intervals%20is%20%3E10%25%20for%20same%20side%2C%20show%20a%20message.%20(That%20is%2C%2010%25%20of%20all%20same%2Dside%20intervals%2C%20combining%20across%20left%2Dleft%20and%20right%2Dright)). I will change to if either is over the threshold

Comment (alexpiet) [reply src/foraging_gui/MyFunctions.py:923]: I suppose its hard to implement things when I tell you inconsistent things. Sorry!

I'm not sure if we want to check if either left or right same side, or combine across sides. But if we do combine, then instead of adding the fractions you should instead concatenate the array of diffs and then compute the fraction:

```
same_side_l = np.diff(left)
same_side_r = np.diff(right)
same_side_combined = np.concat([same_side_l, same_side_r])
same_side_frac = np.mean(same_side_combined <= threshold)
```

If we add the fractions, then the combined fraction ranges from 0 to 2, so that doesn't make sense. 

I guess I would err on the side of checking the combined fraction, since it should be more stable when there are a small number of licks

File: src/foraging_gui/MyFunctions.py
Comment (alexpiet) [reply src/foraging_gui/MyFunctions.py:933]: Ok, I understand!

File: src/foraging_gui/MyFunctions.py
Comment (alexpiet) [reply src/foraging_gui/MyFunctions.py:940]: Ok, I understand whats happening now and it makes sense. I would add a few comments so we can understand how this works in the future. Just explain the structure of `merged_sorted` , and `cross_sides` (which contains only the ITIs for cross side intervals

**Comment by micahwoodard** [comment]
Tested with mocked data. @ellahiltonvano would there be a good mouse to test this feature on? I currently can't test on a rig because the lick times will always be empty. I am fairly confident it will work based on mocked data so if we can't test with a real mouse, I think it will probably be okay

**Review by alexpiet — APPROVED** [review]
Feel free to merge after you finished any testing. A screenshot would be nice before merging.

**Comment by ellahiltonvano** [comment]
@micahwoodard as discussed we could use one of madeline's training mice that start next week since they are not assigned to a particular scientist. If you want to merge before then we can just do normal testing on monday

**Comment by alexpiet** [comment]
> * Will it be visible if the user doesn't click on the lick distribution dialog?
> 
> No. Is that something we want? I think I assumed that the message should be on the lick distribution dialog but should it be a separate window or pop up?

Up to @ellahiltonvano

**Comment by ellahiltonvano** [comment]
@alexpiet I think if we want to use this as a flag that something is going on with the lick detection board then eventually a popup to alert the user would be helpful. This would requires scis defining a very specific value of what is acceptable though

Currently:
ILI > 0.05 is good, no tuning necessary
ILI < 0.05 indicates crosstalk (right to left licks, left to right)
An ILI < 0.05 indicates rebound (left and right licks)

Here is the [SOP](https://alleninstitute.sharepoint.com/:w:/s/NeuralDynamics/EUFfxQ-ah3pOrWvPh_Y5YnoBPr3rGa6UtkP09c5rVo4J0A?e=RPT9fB), for reference. Let me know if y'all think it would be helpful to just code in alarms for these specific abnormalities.

**Comment by micahwoodard** [comment]
@alexpiet, Ella and I chatted about this a little bit. Instead of a pop up that would prompt RA's to alert someone/submit a ticket, would it make sense to log an error which will create an issue if either same side licks or cross side licks exceeds the threshold? This would ensure that these alerts won't get lost or ignored with the other windows. Might take some work making sure it only logs one error and not all instances, but would ensure this issue is being looked at

**Comment by alexpiet** [comment]
> @alexpiet, Ella and I chatted about this a little bit. Instead of a pop up that would prompt RA's to alert someone/submit a ticket, would it make sense to log an error which will create an issue if either same side licks or cross side licks exceeds the threshold? This would ensure that these alerts won't get lost or ignored with the other windows. Might take some work making sure it only logs one error and not all instances, but would ensure this issue is being looked at

That works! I would still add a warning in the warning box (especially if you make a queue log)

**Comment by micahwoodard** [comment]
> That works! I would still add a warning in the warning box (especially if you make a queue log)
The queue log should also display all error log messages so that should be taken care of if implemented

**Comment by micahwoodard** [comment]
@ellahiltonvano is there a mouse that we can test with this week? If so what time/works best for you?

**Comment by ellahiltonvano** [comment]
@micahwoodard we could run one of madeline's training mice on a branch with this update. Wednesday afternoon (12:30) would work!

**Comment by micahwoodard** [comment]
Yep that is perfect thanks Ella!

**Comment by micahwoodard** [comment]
logging error
- add box number 
- mouse
-

**Comment by micahwoodard** [comment]
Tested with Ella and Madeline this afternoon so merging in

**Pull request merged**
